libraries/Microsoft.Bot.Builder.Expressions/BuiltInFunctions.cs

```diff
@@ -902,170 +971,91 @@ private static (object value, string error) AccessIndex(object instance, int ind
-                                    while (list.Count <= iindex.Value)
-                                    {
-                                        list.Add(null);
-                                    }
-                                }
-
-                                // Assign value or expected list size or object
-                                if (list is JArray arr)
-                                {
-                                    result = value != null ? JToken.FromObject(value)
-                                        : (expected.HasValue ? (object)new JArray() : new JObject());
-                                    arr[iindex.Value] = (JToken)result;
-                                }
-                                else
-                                {
-                                    result = value ?? (expected.HasValue ? (object)new List<object>(expected.Value + 1) : (object)new Dictionary<string, object>());
-                                    list[iindex.Value] = result;
-                                }
-                            }
-                            else
-                            {
-                                error = $"{children[0]} is not a list.";
-                            }
+                            (value, error) = AccessProperty(inst, idxStr);
                         }
                         else
                         {
-                            error = $"{children[0]} is not a valid path";
+                            error = $"Could not coerce {index}<{idxValue.GetType()}> to an int or string";
                         }
                     }
                 }
+                return (value, error);
             }
-            else
-            {
-                error = $"{path} is not a path that can be set to a value.";
-            }
-
-            return (result, error);
         }
 
-        private static (object value, string error) SetPathToValue(Expression expr, object state)
+
+        private static (object value, string error) SetPathToValue(Expression expr, IMemoryScope state)
         {
-            var path = expr.Children[0];
-            var valueExpr = expr.Children[1];
-            var (value, error) = valueExpr.TryEvaluate(state);
-            if (error == null)
+            try
             {
-                (_, error) = SetPathToValue(path, null, value, state);
-                if (error != null)
+                var (path, left) = TryAccumulatePath(expr.Children[0], state);
+                if (left != null)
                 {
-                    value = null;
+                    // the expression can't be fully merged as a path
+                    return (null, $"{expr.Children[0].ToString()} is not a valid path to set value");
                 }
+
+                var (value, error) = expr.Children[1].TryEvaluate(state);
+                return (state.SetValue(path, value), null);
+            }
+            catch (Exception e)
```

Comment:
Same here--what is throwing an exception?

Reply:
i removed the exception handling, and let the interface explicit about value vs error.